packages/contracts-bedrock/src/L2/GasTank.sol

```diff
@@ -0,0 +1,119 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.25;
+
+import { IL2ToL2CrossDomainMessenger } from "interfaces/L2/IL2ToL2CrossDomainMessenger.sol";
+import { ICrossL2Inbox, Identifier } from "interfaces/L2/ICrossL2Inbox.sol";
+import { IGasTank } from "interfaces/L2/IGasTank.sol";
+import { Predeploys } from "src/libraries/Predeploys.sol";
+import { SafeSend } from "src/universal/SafeSend.sol";
+
+contract GasTank is IGasTank {
+    uint256 public constant MAX_DEPOSIT = 0.01 ether;
+    uint256 public constant WITHDRAWAL_DELAY = 7 days;
+    // TODO: Calculate claim overhead
+    uint256 public constant CLAIM_OVERHEAD = 100_000;
+
+    IL2ToL2CrossDomainMessenger public constant MESSENGER =
+        IL2ToL2CrossDomainMessenger(Predeploys.L2_TO_L2_CROSS_DOMAIN_MESSENGER);
+
+    mapping(address => uint256) public balanceOf;
+    mapping(address gasProvider => Withdrawal) public withdrawals;
+    mapping(bytes32 => bool) public claimed;
+    mapping(address gasProvider => mapping(bytes32 msgHash => bool)) public flaggedMessages;
+
+    // Deposit funds into the gas tank, from which the relayer can claim the repayment after relaying
+    function deposit(address _to) external payable {
+        uint256 newBalance = balanceOf[_to] + msg.value;
+
+        if (newBalance > MAX_DEPOSIT) revert MaxDepositExceeded();
+
+        balanceOf[_to] = newBalance;
+        emit Deposit(_to, msg.value);
+    }
+
+    function initiateWithdrawal(uint256 amount) external {
+        // Ensure the caller has enough balance
+        if (balanceOf[msg.sender] < amount) revert InsufficientBalance();
+
+        // Record the pending withdrawal
+        withdrawals[msg.sender] = Withdrawal({ timestamp: block.timestamp, amount: amount });
+
+        // Emit an event for the withdrawal initiation
+        emit WithdrawalInitiated(msg.sender, amount);
+    }
+
+    function finalizeWithdrawal(address to) external {
+        Withdrawal memory withdrawal = withdrawals[msg.sender];
+
+        // Ensure the withdraw is not pending
+        if (block.timestamp < withdrawal.timestamp + WITHDRAWAL_DELAY) revert WithdrawPending();
+
+        // Update the balance
+        uint256 amount = balanceOf[msg.sender] < withdrawal.amount ? balanceOf[msg.sender] : withdrawal.amount;
+        balanceOf[msg.sender] -= amount;
+
+        // Clear the pending withdrawal
+        delete withdrawals[msg.sender];
+
+        // Send the funds to the recipient
+        new SafeSend{ value: amount }(payable(to));
+
+        emit WithdrawalFinalized(msg.sender, to, amount);
+    }
+
+    // Flag a message into the gas tank so the relayer is aware of it, and can claim the funds after relaying
+    function flag(bytes32 rootMessageHash) external {
+        flaggedMessages[msg.sender][rootMessageHash] = true;
+        emit Flagged(rootMessageHash);
+    }
+
+    // Claim repayment for a relayed message
+    function claim(Identifier calldata id, address gasProvider, bytes calldata payload) external {
```

Comment:
when the relayer goes to claim by calling this function, how will the relayer know what `gasProvider` to set? what do you think of adding a mapping from `rootMessageHash` to gasProvider(s)?

Reply:
I like that. Another option is to maintain two mappings, that way when the relayer goes to claim it doesn't have to loop over every entry in the mapping to see if that gasProvider has flagged the message:

```solidity
mapping(address gasProvider => mapping(bytes32 msgHash => bool)) public flaggedMessages;
mapping(bytes32 msgHash => address[] gasProviders) public gasProviders;

function flag(bytes32 rootMessageHash) external {
    flaggedMessages[msg.sender][rootMessageHash] = true;
    gasProviders[rootMessageHash].push(msg.sender);
    emit Flagged(rootMessageHash, msg.sender);
}
```

Reply:
another option is we leave this as-is and just have and the relayer be responsible for indexing the `Flagged` event and manages the mapping of `gasProvider` to `rootMessageHash` offchain.
```solidity
mapping(address gasProvider => mapping(bytes32 msgHash => bool)) public flaggedMessages;
```